[DOC] Update TraceQL query editor doc (#91910)
grafana/grafana: docs/sources/datasources/tempo/query-editor/_index.md

You can use these modes by themselves or in combination to create building blocks to generate custom queries.

![The three query types: Search, TraceQL, and Service Graph](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-types.png)

The three **Query types** are:

- The **Search** query builder provides a user interface for building a TraceQL query.
- The **TraceQL** query editor lets you write your own TraceQL query with assistance from autocomplete.
- The **Service Graph** view displays a visual relationship between services. Refer to the [Service graph](ref:service-graph) documentation for more information.

### Search query builder

The Search query builder provides drop-down lists and text fields to help you write a query.
The query builder is ideal for people who aren't familiar with or want to learn TraceQL.

Refer to the [Search using the TraceQL query builder documentation]({{< relref "./traceql-search" >}}) to learn more about creating queries using convenient drop-down menus.

![The Search query builder](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-search-v11.png)

> ### TraceQL Query editor

Author: ```suggestion
### TraceQL query editor
```

> The TraceQL query editor lets you search by trace ID and write TraceQL queries using autocomplete.

Reviewer: ```suggestion
The **TraceQL** query editor lets you search by trace ID and write TraceQL queries using autocomplete.
```

Refer to the [TraceQL query editor documentation]({{< relref "./traceql-editor" >}}) to learn more about constructing queries using a code-editor-like experience.

![The TraceQL query editor](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-traceql-v11.png)

You can also search for a Trace ID by entering a trace ID into the query field.

### Service graph view

> Grafana’s service graph view uses metrics to display span request rates, error rates, and durations, as well as service graphs.

Reviewer: To keep this consistent, gonna format this like the others.
```suggestion
Grafana’s **Service Graph** view uses metrics to display span request rates, error rates, and durations, as well as service graphs.
```

Author: Thank you. So good to have a fresh set of eyes on this.

Once the requirements are set up, this pre-configured view is immediately available.

Using the service graph view, you can:

> - Discover spans which are consistently erroring and the rates at which they occur
> - Get an overview of the overall rate of span calls throughout your services
> - Determine how long the slowest queries in your service take to complete
> - Examine all traces that contain spans of particular interest based on rate, error, and duration values (RED signals)

Reviewer: ```suggestion
- Discover spans which are consistently erroring and the rates at which they occur.
- Get an overview of the overall rate of span calls throughout your services.
- Determine how long the slowest queries in your service take to complete.
- Examine all traces that contain spans of particular interest based on rate, error, and duration values (RED signals).
```

For more information about the service graph, refer to [Service graph](../service-graph/).

![Screenshot of the Service Graph view](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-query-service-graph.png)

## Use TraceQL panels in dashboards

To add TraceQL panels to your dashboard, refer to the [Traces panel documentation](/docs/grafana/latest/panels-visualizations/visualizations/traces/).

To learn more about Grafana dashboards, refer to the [Use dashboards documentation](/docs/grafana/latest/dashboards/use-dashboards/).

## Set options for query builder and editor

The following options are available for the **Search** and **TraceQL** query types.
You can modify these settings in the **Options** section.

![Options section in the query editors](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-options.png)

After changing any option, re-run the query to apply the updates.

Limit
: Determines the maximum number of traces to return. Default value is `20`.

Span Limit
: Sets the maximum number of spans to return for each spanset. Default value is `3`.

Table Format

> : Determines whether the query results table is displayed focused on **Traces** or **Spans**. **Traces** is the default selection. When **Traces** is selected, the results table starts with the Trace ID. When **Spans** is selected, the table starts with the Trace Service.

Reviewer: lowercasing these unless they're UI elements that you're calling out, in which case they should be bold
```suggestion
: Determines whether the query results table is displayed focused on **Traces** or **Spans**. **Traces** is the default selection. When **Traces** is selected, the results table starts with the trace ID. When **Spans** is selected, the table starts with the trace dervice.
```

Step
: Defines the step for metrics quires. Use duration notation, for example, `30ms` or `1m`.

Streaming

> : Indicates if streaming is active. Streaming lets you view partial query results before the entire query completes. Activating Streaming adds the **Table - Streaming Progress** section to the query results.

Reviewer: ```suggestion
: Indicates if streaming is active. Streaming lets you view partial query results before the entire query completes. Activating streaming adds the **Table - Streaming Progress** section to the query results.
```

    - pattern: /docs/grafana/
      destination: /docs/grafana/<GRAFANA_VERSION>/explore/query-management/
    - pattern: /docs/grafana-cloud/
      destination: /docs/grafana/<GRAFANA_VERSION>/explore/query-management/
  query-inspector:
    - pattern: /docs/grafana/
      destination: /docs/grafana/<GRAFANA_VERSION>/explore/explore-inspector/
    - pattern: /docs/grafana-cloud/
      destination: /docs/grafana/<GRAFANA_VERSION>/explore/explore-inspector/
---

# Query tracing data

The Tempo data source's query editor helps you query and display traces from Tempo in [Explore](ref:explore).
The queries use [TraceQL](/docs/tempo/latest/traceql), the query language designed specifically for tracing.

For general documentation on querying data sources in Grafana, refer to [Query and transform data](ref:query-transform-data).

## Before you begin

You can compose TraceQL queries in Grafana and Grafana Cloud using **Explore** and a Tempo data source.

### TraceQL knowledge helpful, but not required

You don't have to know TraceQL to create a query.
You can use the **Search** query builder's user interface to select options to search your data.
These selections generate a TraceQL query.

> Any query generated in Build mode can be transferred to the Code mode, where you can edit the query directly.

Author: I'm glad you caught this because Tempo data source doesn't actually have a build and code mode. This is the wrong terminology. I started updating the content bcause the videos on the Grafana YouTube channel use this terminology. Loki has Builder and Code mode, but Tempo doesn't.

> Any query generated in Build mode can be transferred to the Code mode, where you can edit the query directly.

Author: ```suggestion
Any query generated using **Search** query builder can be transferred to the **TraceQL** query editor, where you can edit the query directly.
```

> Any query generated in Build mode can be transferred to the Code mode, where you can edit the query directly.

Author: I reworded the section so mode should be removed.

To learn more about how to query by TraceQL, refer to the [TraceQL documentation](/docs/tempo/latest/traceql).

## Choose a query editing mode

> The query editor has three modes, or **Query types**, that you can use to explore your tracing data.

Author: This was also a left over from a rewrite I reverted. I'm glad you're reading this so it can be clarified. 
```suggestion
The query editor has three **Query types** that you can use to explore your tracing data.
```

> You can use these modes by themselves or in combination to create building blocks to generate custom queries.

Author: ```suggestion
You can use these modes by themselves or in combination to create building blocks to generate custom queries.
Adding another query adds a new query block. 
Refer to [Use query types together](#use-query-types-together) for more information.
```

## Set options for query builder and editor

The following options are available for the **Search** and **TraceQL** query types.
You can modify these settings in the **Options** section.

![Options section in the query editors](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-options.png)

After changing any option, re-run the query to apply the updates.

Limit
: Determines the maximum number of traces to return. Default value is `20`.

Span Limit
: Sets the maximum number of spans to return for each spanset. Default value is `3`.

Table Format
: Determines whether the query results table is displayed focused on **Traces** or **Spans**. **Traces** is the default selection. When **Traces** is selected, the results table starts with the trace ID. When **Spans** is selected, the table starts with the trace service.

Step
: Defines the step for metrics quires. Use duration notation, for example, `30ms` or `1m`.

Streaming
: Indicates if streaming is active. Streaming lets you view partial query results before the entire query completes. Activating streaming adds the **Table - Streaming Progress** section to the query results.

## Use query types together

> You can use **+ Add query** to create customized queries that use one or more of the query types together.

Author: ```suggestion
You can use **+ Add query** to create customized queries that use one or more of the query types together.
Each time you add a new query, it adds a new section, or query block, that contains **Search**, **TraceQL**, or **Service Graph** user interface. 
```

> Added query and results table appear in the navigation under **Queries** and **Tables** respectively.

Author: I've added a note about query blocks at line 74 and I've also added more information in this section.

You can use the navigation to view query, results table, and service graph blocks.

{{< video-embed src="/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-editor.mp4" max-width="800px" class="my-cool-video" caption="Navigating through the query blocks" align="center" >}}

To add a query block:

1. Select **+ Add query**.
1. Choose a query type: **Search**, **TraceQL**, or **Service Graph**.

To remove a query block, select the **Remove query** trash can icon.

To rename a block, select the **Rename** edit icon next to the query block name.
The name changes in the queries and table list.

### Additional query block options

Each query block has a set of icons in the right top corner.

![The additional options toolbar](/media/docs/grafana/data-sources/tempo/query-editor/tempo-ds-additional-options-toolbar.png)

These icons include:

Show data source help

> : Displays the **Tempo Cheat Sheet** with links to documentation.

Author: The screen name that shows is called the Tempo Cheat Sheet, so does it then count as a UI element? 
![image](https://github.com/user-attachments/assets/4b91d40d-4412-4dc5-a86d-beaaf08b2079)

> : Displays the **Tempo Cheat Sheet** with links to documentation.

Reviewer: Ah okay, I didn't see that. Yep, UI element!

Create recorded query
: Lets you save the current query block as a recorded query. This option is available in Grafana Cloud and Grafana Enterprise. For more information, refer to [Recorded queries](ref:recorded-queries).

Duplicate query
: Copies the current block and adds a new identical block.

Remove query
: Deletes the query block.

### Use query history and query inspector

> **Explore** provides a history of all queries you've used within a data source and an inspector that lets you view stats, inspect queries, view JSON, and general data for your data source queries.

Reviewer: ```suggestion
**Explore** provides a history of all queries you've used within a data source and an inspector that lets you view stats, inspect queries, view JSON, and general information for your data source queries.
```

For more information, refer to the [Query inspector in Explore](ref:query-inspector) and [Query management in Explore](ref:query-history-management) documentation.

## Cross-tenant TraceQL queries

> If you have configured multi-stack Tempo data source, you can perform TraceQL queries across those stacks and tenants.

Reviewer: ```suggestion
If you've configured a multi-stack Tempo data source, you can perform TraceQL queries across those stacks and tenants.
```

Queries performed using the cross-tenant configured data source, in either **Explore** or inside of dashboards,
are performed across all the tenants that you specified in the **X-Scope-OrgID** header.

<!-- vale Grafana.Spelling = NO -->

TraceQL queries that compare multiple spansets may not correctly return all traces in a cross-tenant query. For instance,

<!-- vale Grafana.Quotes = YES -->

```
{ span.attr1 = "bar" } && { span.attr2 = "foo" }
```

TraceQL evaluates a contiguously stored trace.
If these two conditions are satisfied in separate tenants, then Tempo doesn't return the trace.

Refer to [Set up a multi-stack Tempo data source in Grafana](https://grafana.com/docs/grafana-cloud/connect-externally-hosted/multi-stack-data-sources/#set-up-a-multi-stack-tempo-data-source-in-grafana) for information about configuring the Tempo data source.

> For information about Tempo configuration requirements, refer to the [Cross-tenant query](https://grafana.com/docs/tempo/<TEMPO_VERSION>/operations/cross_tenant_query/) and [Enable multitenancy](https://grafana.com/docs/tempo/<TEMPO_VERSION/operations/multitenancy/) documentation.

Reviewer: ```suggestion
For information about Tempo configuration requirements, refer to the [Cross-tenant query](https://grafana.com/docs/tempo/<TEMPO_VERSION>/operations/cross_tenant_query/) and [Enable multitenancy](https://grafana.com/docs/tempo/<TEMPO_VERSION>/operations/multitenancy/) documentation.
```
